packages/server/src/__tests__/plugin/drainHttpServer/stoppable.test.ts

```diff
@@ -211,17 +246,17 @@ Object.keys(schemes).forEach((schemeName) => {
       ]);
       const start = Date.now();
       const closeEventPromise = a.event(server, 'close');
-      const gracefully = await stopper.stop(500);
+      const [gracefully] = await Promise.all([
+        stopper.stop(500),
+        jest.advanceTimersByTime(500),
+      ]);
       await closeEventPromise;
-      // These tests are a bit flakey; we should figure out a way to usefully
-      // test the grace period behavior without leading to flakiness due to
-      // speed variation.
+
       const elapsed = Date.now() - start;
-      expect(elapsed).toBeGreaterThanOrEqual(450);
-      expect(elapsed).toBeLessThanOrEqual(550);
+      expect(elapsed).toBe(500);
```

Comment:
I might be wrong but I think this isn't actually testing anything any more.

Originally, what the test showed was that in this case (where there are a couple requests "hanging" because the response hasn't "end"ed), it would *not* immediately kill the requests, but that we would in fact get to wait 500 ms.  So this could fail by getting a really small value for `elapsed` if it didn't actually wait the whole grace period.

But in this case, if I understand correctly, this is merely just testing that the `jest.advanceTimersByTime` ran.  So it can't really fail?

Here's an idea.  Stopper is a fully internal class so we can do whatever we want with it. What if we replaced `stopGracePeriodMillis` with an AbortSignal (we even have a dependency on `node-abort-controller` as of Monday...).  So the `setTimeout` would occur in the `drainServer` hook in ApolloServerPluginDrainHttpServer but Stopper wouldn't care about time any more (it would just do a `signal.on('abort')` or whatever instead of a `setTimeout`).

Then the tests wouldn't have to mock out time at all: they would be able to deterministically show that `stopper.stop(signal)` doesn't return at all until the test calls `controller.abort()`.

(Other tests have similar concerns.)

Reply:
(Note that there is a test that uses an external process; for this one you could send a SIGUSR1 or something to trigger the abortcontroller?)

Reply:
I implemented this suggestion in https://github.com/apollographql/apollo-server/pull/7232